File: javascript_version/array_and_strings/uncompress/uncompress.js
// uncompress
// Write a function, uncompress, that takes in a string as an argument. The input string 
// will be formatted into multiple groups according to the following pattern:

// <number><char>

// for example, '2c' or '3a'.
// The function should return an uncompressed version of the string where each 'char' of a group is repeated 'number' times consecutively. 
// You may assume that the input string is well-formed according to the previously mentioned pattern.


//more optimized

const uncompress = (s) => {

    let i = 0;
    let j = 0;
    let numbers = "0123456789";
    let result = "";
    while (j <= s.length) {
        if (numbers.includes(s[j])) {
            j++;
        }
        else {
            let cut = parseInt(s.slice(i, j));
            for (let x = 0; x < cut; x++) {
                result += s[j];
            }
            j++;
            i = j;
        }

    }
    return result;
};

console.log(uncompress("2c3a1t")); // -> 'ccaaat'
console.log(uncompress("4s2b")); // -> 'ssssbb'
console.log(uncompress("2p1o5p")); // -> 'ppoppppp'
console.log(uncompress("3n12e2z")); // -> 'nnneeeeeeeeeeeezz'
console.log(uncompress("127y")); // ->'yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy'

File: cpp_version/array_and_strings/uncompress/uncompress.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// uncompress
// Write a function, uncompress, that takes in a string as an argument. The input string
// will be formatted into multiple groups according to the following pattern:

// <number><char>

// for example, '2c' or '3a'.
// The function should return an uncompressed version of the string where each 'char' of a group is repeated 'number' times consecutively.
// You may assume that the input string is well-formed according to the previously mentioned pattern.

std::string uncompress(std::string s) {
    // todo

    std::string result = "";

    return result;
}

int main() {

    std::cout << uncompress("2c3a1t") << std::endl;  // -> "ccaaat"
    std::cout << uncompress("4s2b") << std::endl;    // -> "ssssbb"
    std::cout << uncompress("2p1o5p") << std::endl;  // -> "ppoppppp"
    std::cout << uncompress("3n12e2z") << std::endl; // -> "nnneeeeeeeeeeeezz"
    std::cout << uncompress("127y") << std::endl;
    // ->"yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy"

    return 0;
}